[visionOS] Move shared simulation connection from the Model process to the GPU process

achan00 opened the pull request:
#### 8746ea87f06c2e65ea34055e95bcb7cbfa1ea59e
<pre>
[visionOS] Move shared simulation connection from the Model process to the GPU process
<a href="https://bugs.webkit.org/show_bug.cgi?id=286366">https://bugs.webkit.org/show_bug.cgi?id=286366</a>
<a href="https://rdar.apple.com/138447079">rdar://138447079</a>

Reviewed by Per Arne Vollan.

Instead of connecting to shared simulation as part of WKREEngine initialization
in the Model process, the Model process will send a message to the UI process to
request for that connection. UI process delegates that request to the GPU process
and returns the resulting connection back to the Model process. Sandbox rules and
entitlements related to the shared simulation connection have been removed from
the Model process, and added to the GPU process instead (only for visionOS).

* Source/WebKit/Configurations/GPUService.xcconfig:
Update the GPU process simulator entitlements for visionOS
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
Add a new message for handling shared simulation connection
* Source/WebKit/GPUProcess/cocoa/GPUProcessCocoa.mm:
(WebKit::GPUProcess::requestSharedSimulationConnection):
Get the shared simulation connection via WKSharedSimulationConnectionHelper using
the model process&apos; audit token.
* Source/WebKit/ModelProcess/ModelProcess.cpp:
(WebKit::ModelProcess::createModelConnectionToWebProcess):
Initialize the WKREEngine with a function to get the shared simulation
connection via a message to the UI process.
(WebKit::ModelProcess::requestSharedSimulationConnection):
* Source/WebKit/ModelProcess/ModelProcess.h:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.GPU.sb.in:
A couple of additions to the GPU process sandbox in order to connect to shared
simulation and to make shareable file handles.
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in:
Remove the mach-lookup rule needed to connect to shared simulation.
* Source/WebKit/Resources/ios/GPUService-visionOS-simulator.entitlements: Copied from Source/WebKit/Resources/ios/ModelService-embedded-simulator.entitlements.
* Source/WebKit/Resources/ios/ModelService-embedded-simulator.entitlements:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::requestSharedSimulationConnection):
Send the shared simulation connection request to the GPU process
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::processWillShutDown):
Reset m_didInitializeSharedSimulationConnection to false.
* Source/WebKit/UIProcess/Model/ModelProcessProxy.h:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.messages.in:
* Source/WebKit/UIProcess/Model/cocoa/ModelProcessProxyCocoa.mm: Added.
(WebKit::ModelProcessProxy::requestSharedSimulationConnection):
Add message checks to make sure the web content process that creates the models
to be loaded by the Model process have the expected model related preferences
set, and that this is the first time this Model process is requesting for the connection.
If all the checks pass, delegate the connection request to GPUProcessProxy.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/289447@main">https://commits.webkit.org/289447@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c41f933cb15130584e94440cfbb6185894f8062d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86808 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6315 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41148 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91655 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37540 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6583 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14376 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67098 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24871 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89811 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5007 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78568 "Found 1 new API test failure: TestWebKitAPI.ContentRuleList.ResourceTypes (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47417 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/86306 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4792 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36658 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75297 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33810 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93549 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13961 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10122 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75900 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14162 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74414 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75095 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18503 "") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19412 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17826 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/6754 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13553 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13984 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19244 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13722 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17167 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15507 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gavin-apple commented on Source/WebKit/Scripts/process-entitlements.sh:
```diff
@@ -536,10 +536,18 @@ fi
     plistbuddy Add :com.apple.security.exception.mach-lookup.global-name array
     plistbuddy Add :com.apple.security.exception.mach-lookup.global-name:0 string com.apple.systemstatus.activityattribution
+if [[ "${WK_PLATFORM_NAME}" == xros ]]; then
+    plistbuddy Add :com.apple.security.exception.mach-lookup.global-name:1 string com.apple.surfboard.applicationservice
```
Is the mach-lookup exception necessary here given it's been added to the GPUP sandbox profile above?

achan00 replied:
Good catch! This has been removed.

mwyrzykowski commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.GPU.sb.in:
```diff
@@ -772,6 +772,9 @@
         SYS_fcntl_nocancel
         SYS_fileport_makefd
+#if PLATFORM(VISION) && ENABLE(MODEL_PROCESS)
+        SYS_fileport_makeport
```
Change looks good to me, just want to double check with @pvollan , @smfr , or @hortont424 on the sandbox change

pvollan replied:
I think it's OK to add this syscall in the GPU process sandbox on visionOS.

pvollan commented on Source/WebKit/Resources/ios/GPUService-visionOS-simulator.entitlements:
```diff
@@ -0,0 +1,16 @@
+	<key>com.apple.private.webkit.use-xpc-endpoint</key>
+	<true/>
+	<key>com.apple.private.webkit.webpush</key>
+	<true/>
```
Is this entitlement required?

achan00 replied:
I started with a copy of XPCService-embedded-simulator.entitlements (as this was what the GPU process was using) and this was in the original. If we are sure we don't need this for the GPU process, we can remove it.

achan00 replied:
Double-checked with Brady that this is not needed for the GPU process. It has been removed.

pvollan commented on Source/WebKit/ModelProcess/ModelProcess.cpp:
```diff
@@ -81,6 +86,18 @@ void ModelProcess::createModelConnectionToWebProcess(WebCore::ProcessIdentifier
         return;
 
+#if PLATFORM(VISION) && ENABLE(GPU_PROCESS)
+    WKREEngine::shared().initializeWithSharedSimulationConnectionGetterIfNeeded([identifier, weakThis = WeakPtr { *this }] (CompletionHandler<void(std::optional<IPC::SharedFileHandle>)>&& completionHandler) {
```
Is `initializeWithSharedSimulationConnectionGetterIfNeeded` a new method?

achan00 replied:
Yes, being introduced with the fix for rdar://142698804.

pvollan reviewed (approved):
R=me.

achan00 committed:
[visionOS] Move shared simulation connection from the Model process to the GPU process
https://bugs.webkit.org/show_bug.cgi?id=286366
rdar://138447079

Reviewed by Per Arne Vollan.

Instead of connecting to shared simulation as part of WKREEngine initialization
in the Model process, the Model process will send a message to the UI process to
request for that connection. UI process delegates that request to the GPU process
and returns the resulting connection back to the Model process. Sandbox rules and
entitlements related to the shared simulation connection have been removed from
the Model process, and added to the GPU process instead (only for visionOS).

* Source/WebKit/Configurations/GPUService.xcconfig:
Update the GPU process simulator entitlements for visionOS
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
Add a new message for handling shared simulation connection
* Source/WebKit/GPUProcess/cocoa/GPUProcessCocoa.mm:
(WebKit::GPUProcess::requestSharedSimulationConnection):
Get the shared simulation connection via WKSharedSimulationConnectionHelper using
the model process' audit token.
* Source/WebKit/ModelProcess/ModelProcess.cpp:
(WebKit::ModelProcess::createModelConnectionToWebProcess):
Initialize the WKREEngine with a function to get the shared simulation
connection via a message to the UI process.
(WebKit::ModelProcess::requestSharedSimulationConnection):
* Source/WebKit/ModelProcess/ModelProcess.h:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.GPU.sb.in:
A couple of additions to the GPU process sandbox in order to connect to shared
simulation and to make shareable file handles.
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in:
Remove the mach-lookup rule needed to connect to shared simulation.
* Source/WebKit/Resources/ios/GPUService-visionOS-simulator.entitlements: Copied from Source/WebKit/Resources/ios/ModelService-embedded-simulator.entitlements.
* Source/WebKit/Resources/ios/ModelService-embedded-simulator.entitlements:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::requestSharedSimulationConnection):
Send the shared simulation connection request to the GPU process
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::processWillShutDown):
Reset m_didInitializeSharedSimulationConnection to false.
* Source/WebKit/UIProcess/Model/ModelProcessProxy.h:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.messages.in:
* Source/WebKit/UIProcess/Model/cocoa/ModelProcessProxyCocoa.mm: Added.
(WebKit::ModelProcessProxy::requestSharedSimulationConnection):
Add message checks to make sure the web content process that creates the models
to be loaded by the Model process have the expected model related preferences
set, and that this is the first time this Model process is requesting for the connection.
If all the checks pass, delegate the connection request to GPUProcessProxy.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/289447@main
